java-frontend/src/main/java/org/sonar/java/resolve/TypeInferenceSolver.java

```diff
@@ -0,0 +1,286 @@
+      return applySubstitution((JavaType.WildCardType) type, substitution);
+    }
+    if (isArrayType(type)) {
+      return applySubstitution((JavaType.ArrayJavaType) type, substitution);
+    }
+    return type;
+  }
+
+  private static boolean isParametrizedType(JavaType type) {
+    return type instanceof JavaType.ParametrizedTypeJavaType;
+  }
+
+  private static boolean isWildcardType(JavaType type) {
+    return type instanceof JavaType.WildCardType;
+  }
+
+  private static boolean isArrayType(JavaType type) {
+    return type instanceof JavaType.ArrayJavaType;
+  }
+
+  private JavaType applySubstitution(JavaType.ParametrizedTypeJavaType type, TypeSubstitution substitution) {
+    TypeSubstitution newSubstitution = new TypeSubstitution();
+    for (Map.Entry<JavaType.TypeVariableJavaType, JavaType> entry : type.typeSubstitution.substitutionEntries()) {
+      newSubstitution.add(entry.getKey(), applySubstitution(entry.getValue(), substitution));
+    }
+    return parametrizedTypeCache.getParametrizedTypeType(type.rawType.getSymbol(), newSubstitution);
+  }
+
+  private JavaType applySubstitution(JavaType.WildCardType wildcard, TypeSubstitution substitution) {
+    JavaType substitutedType = substitution.substitutedType(wildcard.bound);
+    if (substitutedType != null) {
+      return parametrizedTypeCache.getWildcardType(substitutedType, wildcard.boundType);
+    }
+    return wildcard;
+  }
+
+  private JavaType applySubstitution(JavaType.ArrayJavaType arrayType, TypeSubstitution substitution) {
+    JavaType rootElementType = arrayType.elementType;
+    int nbDimensions = 1;
+    while (rootElementType.isTagged(JavaType.ARRAY)) {
+      rootElementType = ((JavaType.ArrayJavaType) rootElementType).elementType;
+      nbDimensions++;
+    }
+    JavaType substitutedType = substitution.substitutedType(rootElementType);
+    if (substitutedType != null) {
+      // FIXME SONARJAVA-1574 a new array type should not be created but reused if already existing for the current element type
+      for (int i = 0; i < nbDimensions; i++) {
+        substitutedType = new JavaType.ArrayJavaType(substitutedType, symbols.arrayClass);
+      }
+      return substitutedType;
+    }
+    return arrayType;
+  }
+
+  TypeSubstitution getSubstitutionFromTypeParams(List<JavaType.TypeVariableJavaType> typeVariableTypes, List<JavaType> typeParams) {
+    TypeSubstitution substitution = new TypeSubstitution();
+    if (typeVariableTypes.size() == typeParams.size()) {
+      // create naive substitution
+      for (int i = 0; i < typeVariableTypes.size(); i++) {
+        JavaType.TypeVariableJavaType typeVariableType = typeVariableTypes.get(i);
+        JavaType typeParam = typeParams.get(i);
+        substitution.add(typeVariableType, typeParam);
+      }
+    }
+    return substitution;
+  }
+
+  private TypeSubstitution getSubstitutionFromArguments(JavaSymbol.MethodJavaSymbol method, List<JavaType> formals, List<JavaType> argTypes) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) The Cyclomatic Complexity of this method "getSubstitutionFromArguments" is 24 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)